File: common/src/main/java/dev/mayaqq/estrogen/client/cosmetics/ui/widgets/CosmeticPreview.java
package dev.mayaqq.estrogen.client.cosmetics.ui.widgets;

import com.teamresourceful.resourcefullib.client.components.CursorWidget;
import com.teamresourceful.resourcefullib.client.screens.CursorScreen;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.GuiGraphics;
import net.minecraft.client.gui.components.AbstractWidget;
import net.minecraft.client.gui.narration.NarrationElementOutput;
import net.minecraft.client.gui.screens.inventory.InventoryScreen;
import net.minecraft.network.chat.CommonComponents;
import net.minecraft.util.Mth;
import net.minecraft.world.entity.player.Player;
import org.jetbrains.annotations.NotNull;
import org.joml.Quaternionf;

public class CosmeticPreview extends AbstractWidget implements CursorWidget {

    private final Player player;
    private boolean dragging = false;
    private float rotation = Mth.PI / 4;

    public CosmeticPreview(int x, int y, int width, int height) {
        super(x, y, width, height, CommonComponents.EMPTY);

        this.player = Minecraft.getInstance().player;
    }

    @Override
    protected void renderWidget(@NotNull GuiGraphics graphics, int mouseX, int mouseY, float partialTicks) {
        if (this.player != null) {
            Quaternionf quaternion = new Quaternionf().rotateZ(Mth.PI).rotateY(this.rotation);
            float yHeadRot = this.player.yBodyRot;
            float yRot = this.player.getYRot();
            float xRot = this.player.getXRot();
            float yHeadRotO = this.player.yHeadRotO;
            float yBodyRot = this.player.yHeadRot;
            this.player.yBodyRot = 180.0F;
            this.player.setYRot(180.0F);
            this.player.setXRot(0f);
            this.player.yHeadRot = this.player.getYRot();
            this.player.yHeadRotO = this.player.getYRot();
            InventoryScreen.renderEntityInInventory(
                    graphics,
                    (int) (getX() + getWidth() / 2f), getY() + getHeight() - 20,
                    (int) (getHeight() / 2.5F), quaternion, null, this.player
            );
            this.player.yBodyRot = yHeadRot;
            this.player.setYRot(yRot);
            this.player.setXRot(xRot);
            this.player.yHeadRot = yHeadRotO;
            this.player.yHeadRotO = yBodyRot;
        }
    }

    @Override
    public boolean mouseClicked(double mouseX, double mouseY, int button) {
        if (isMouseOver(mouseX, mouseY) && isValidClickButton(button)) {
            this.dragging = true;
            return true;
        }
        return false;
    }

    @Override
    public boolean mouseReleased(double mouseX, double mouseY, int button) {
        this.dragging = false;
        return super.mouseReleased(mouseX, mouseY, button);
    }

    @Override
    public boolean mouseDragged(double d, double e, int i, double f, double g) {
        this.rotation += (float) f * 0.15F;
        return true;
    }

    @Override
    protected void updateWidgetNarration(@NotNull NarrationElementOutput narrationElementOutput) {

    }

    @Override
    public CursorScreen.Cursor getCursor() {
        return this.dragging ? CursorScreen.Cursor.RESIZE_EW : CursorScreen.Cursor.DEFAULT;
    }
}

File: src/main/kotlin/dev/mayaqq/estrogen/client/content/screen/cosmetics/widget/CosmeticPreview.kt
package dev.mayaqq.estrogen.client.content.screen.cosmetics.widget

import dev.mayaqq.cynosure.helpers.McFont
import dev.mayaqq.cynosure.text.CommonText
import dev.mayaqq.cynosure.text.unaryMinus
import dev.mayaqq.estrogen.client.content.screen.EstrogenButton
import dev.mayaqq.estrogen.client.content.screen.EstrogenMenuScreen
import net.minecraft.client.gui.GuiGraphics
import net.minecraft.client.gui.screens.inventory.InventoryScreen
import net.minecraft.util.Mth
import net.minecraft.world.entity.player.Player
import org.joml.Quaternionf
import org.joml.Vector3f

class CosmeticPreview(val player: Player?, x: Int, y: Int, width: Int, height: Int) : EstrogenButton(
    x, y, width, height,
    arrayOf(),
    OnPress {},
    DEFAULT_NARRATION,
    EstrogenMenuScreen.transBlue,
    false, true
) {
    override fun renderWidget(graphics: GuiGraphics, mouseX: Int, mouseY: Int, partialTick: Float) {
        super.renderWidget(graphics, mouseX, mouseY, partialTick)

        if (this.message == CommonText.EMPTY) this.message = -"gui.estrogen.cosmetics.no_preview"
        if (player != null) {

            val quaternion = Quaternionf().rotateZYX(Mth.PI, yRotation, 0F)
            val yHeadRot: Float = player.yBodyRot
            val yRot: Float = player.yRot
            val xRot: Float = player.xRot
            val yHeadRotO: Float = player.yHeadRotO
            val yBodyRot: Float = player.yHeadRot
            player.yBodyRot = 180.0f
            player.yRot = 180.0f
            player.xRot = 0f
            player.yHeadRot = player.yRot
            player.yHeadRotO = player.yRot
            InventoryScreen.renderEntityInInventory(
                graphics,
                (x + getWidth() / 2f),
                y + getHeight() - 20f,
                getHeight() / 2.5f,
                Vector3f(),
                quaternion,
                null,
                player
            )
            player.yBodyRot = yHeadRot
            player.yRot = yRot
            player.xRot = xRot
            player.yHeadRot = yHeadRotO
            player.yHeadRotO = yBodyRot
        } else {
            this.renderString(graphics, McFont, 0xFFFFFF)
        }
    }

    override fun mouseClicked(mouseX: Double, mouseY: Double, button: Int): Boolean = isMouseOver(mouseX, mouseY) && isValidClickButton(button)


    override fun mouseDragged(mouseX: Double, mouseY: Double, button: Int, dragX: Double, dragY: Double): Boolean {
        yRotation += dragX.toFloat() * 0.15f
        return true
    }

    companion object {
        private var yRotation = Mth.PI / 4
    }
}